core/src/main/java/org/elasticsearch/transport/netty/MessageChannelHandler.java

```diff
@@ -342,4 +313,66 @@ public void onFailure(Throwable e) {
             }
         }
     }
+
+    /**
+     * Internal helper class to store characteristic offsets of a buffer during processing
+     */
+    protected static final class Marker {
+        private final ChannelBuffer buffer;
+        private final int size;
+        private final int markedReaderIndex;
+        private final int expectedReaderIndex;
+
+        public Marker(ChannelBuffer buffer) {
+            this.buffer = buffer;
+            this.size = buffer.getInt(buffer.readerIndex() - 4);
```

Comment:
I looked whether it makes sense to remove the class and replace it with validation methods. As this class is stateful, we'd need to pass additional parameters to the validation methods which would lead to method signatures with 6 (for requests) and 7 parameters (for responses). I have added constants for all parts of the header and eliminated all magic numbers related to these headers.